muffnn/mlp/tests/util.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+from scipy.stats import pearsonr
+from sklearn.model_selection import train_test_split
+
+
+def assert_sample_weights_work(make_dataset_func, dataset_kwargs,
+                               make_classifier_func):
+    """Ensure we handle sample weights for the given classifier.
+
+    Ensures sample weights are taking affect by training multiple
+    classifiers on different permutations of the same data with
+    sample weights balancing the differences and by ensuring that
+    when trying to learn two datasets simultaneously, the estimator
+    performs best on the dataset weighted most heavily.
+
+    Parameters
+    ----------
+    make_dataset_func : function
+        Function that when called with ``dataset_kwargs`` will
+        return an X and y for testing the given estimator.
+    dataset_kwargs : dict
+        Arguments that will be passed to ``make_dataset_func``
+        when invoked.
+    make_classifier_func : function
+        Function that when called returns the estimator to be tested.
+    """
+    X, y = make_dataset_func(random_state=42, **dataset_kwargs)
+    X2, y2 = make_dataset_func(random_state=24, **dataset_kwargs)
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.33, random_state=42)
+
+    rng = np.random.RandomState(42)
+
+    def assert_clfs_almost_equal(clf1, clf2):
+        """Test outputs are approximately equal."""
+        if hasattr(clf1, 'predict_proba'):
+            disagreement = (clf1.predict_proba(X_test) -
+                            clf2.predict_proba(X_test))
+            assert np.abs(disagreement).mean() < 0.05
+        else:
+            correlation, _ = pearsonr(clf1.predict(X_test),
+                                      clf2.predict(X_test))
+            assert correlation > .98
+
+    # A classifier with no sample weights and a classifier with
+    # sample weights of all 1 should be equivalent.
+    sample_weight = np.ones(X_train.shape[0])
+    clf1 = make_classifier_func().fit(X_train, y_train)
+    clf2 = make_classifier_func().fit(
+        X_train, y_train,
+        sample_weight=sample_weight,
+    )
+    assert_clfs_almost_equal(clf1, clf2)
+
+    # Fitting twice with half sample-weights should result
+    # in same classifier as fitting once with full weights.
```

Comment:
hmm... i don't think running for twice as many epochs with sample weights scaled by 0.5 will lead to the same model for non-vanilla SGD.  Also, shouldn't the models be the same after a single fit of clf2 because of the batch sample weight normalization?

Does this fail if partial_fit is only called once below for clf2?

Reply:
Yeah, I don't think this test makes sense because we're normalizing by the sum of the weights.
My original thinking was to shrink the loss with the weights to force the training to take smaller steps, but normalization means this isn't happening -- I'll remove this test.